### PR TITLE
[DLS-6490] DES to IF migration of API#1303 

### PR DESCRIPTION
**Pull request by ruthdas03**

### CODE REVIEW

File: app/uk/gov/hmrc/helptosave/connectors/IFConnector.scala [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+  val payeURL: String = servicesConfig.baseUrl("paye-personal-details")
+
+  implicit val booleanShow: Show[Boolean] = Show.show(if (_) "Y" else "N")
+
+  val originatorIdHeader: (String, String) = "Originator-Id" -> servicesConfig.getString(
+    "microservice.services.paye-personal-details.originatorId")
+
+  def payePersonalDetailsUrl(nino: String): String = s"$payeURL/pay-as-you-earn/02.00.00/individuals/$nino"
+
+  override def getPersonalDetails(nino: NINO)(implicit hc: HeaderCarrier, ec: ExecutionContext): Future[HttpResponse] =
```
Comment (johno1985) [inline app/uk/gov/hmrc/helptosave/connectors/IFConnector.scala:50]: execution context should be at the class level

File: app/uk/gov/hmrc/helptosave/services/HelpToSaveService.scala [hunk 2]
```diff
@@ -136,54 +138,105 @@ class HelpToSaveServiceImpl @Inject()(
   def getPersonalDetails(nino: NINO)(implicit hc: HeaderCarrier, ec: ExecutionContext): Result[PayePersonalDetails] =
     EitherT({
       val timerContext = metrics.payePersonalDetailsTimer.time()
+      if(appConfig.isFeatureSwitchEnabled(CallDES)){
```
Comment (johno1985) [inline app/uk/gov/hmrc/helptosave/services/HelpToSaveService.scala]: please can this be refactored? there is a lot of repeated code here

File: app/uk/gov/hmrc/helptosave/config/featureSwitches/FeatureSwitching.scala [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+
+import play.api.Configuration
+import uk.gov.hmrc.helptosave.config.featureSwitches.FeatureSwitch.FeatureSwitch
+
+trait FeatureSwitching {
+  implicit val config: Configuration
+  val FEATURE_SWITCH_ON = "true"
+  val FEATURE_SWITCH_OFF = "false"
+
+  def enableFeatureSwitch(featureSwitch: FeatureSwitch): Unit =
```
Comment (johno1985) [inline app/uk/gov/hmrc/helptosave/config/featureSwitches/FeatureSwitching.scala:27]: i don't think we need these, can just use mocking in unit tests instead

**Review by johno1985 — CHANGES_REQUESTED** [review]
a few comments added, plus, please correct PR title

File: app/uk/gov/hmrc/helptosave/config/featureSwitches/FeatureSwitch.scala [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package uk.gov.hmrc.helptosave.config.featureSwitches
+
+object FeatureSwitch {
```
Comment (johno1985) [inline app/uk/gov/hmrc/helptosave/config/featureSwitches/FeatureSwitch.scala:19]: a flat boolean config value will suffice

**Review by johno1985 — CHANGES_REQUESTED** [review]

File: app/uk/gov/hmrc/helptosave/services/HelpToSaveService.scala [hunk 5]
```diff
@@ -55,23 +59,24 @@ trait HelpToSaveService {
...
+                                       ucThresholdProvider: ThresholdManagerProvider)(
+                                       implicit ninoLogMessageTransformer: LogMessageTransformer,
+                                       appConfig: AppConfig, val config: Configuration)
+  extends HelpToSaveService with Logging with FeatureSwitching {
 
   override def getEligibility(nino: NINO, path: String)(
     implicit hc: HeaderCarrier,
     ec: ExecutionContext): Result[EligibilityCheckResponse] =
     for {
-      threshold  <- EitherT.liftF(getThresholdValue())
```
Comment (johno1985) [inline app/uk/gov/hmrc/helptosave/services/HelpToSaveService.scala:72]: in the future, please can we keep formatting changes in a separate commit?

**Review by johno1985 — CHANGES_REQUESTED** [review]

File: app/uk/gov/hmrc/helptosave/config/AppConfig.scala [hunk 6]
```diff
@@ -29,18 +29,24 @@ import scala.jdk.CollectionConverters._
 
 @Singleton
 class AppConfig @Inject()(
+  val config: Configuration,
```
Comment (johno1985) [inline app/uk/gov/hmrc/helptosave/config/AppConfig.scala:32]: not needed, please remove

File: app/uk/gov/hmrc/helptosave/connectors/DESConnector.scala [hunk 7]
```diff
@@ -31,7 +31,7 @@ import scala.concurrent.{ExecutionContext, Future}
 
 @ImplementedBy(classOf[DESConnectorImpl])
 trait DESConnector {
-  def desCorrelationId(response: HttpResponse): String = response.header("CorrelationId").getOrElse("-")
+  def desCorrelationId(response: HttpResponse): String = response.header("DesCorrelationId").getOrElse("-")
```
Comment (johno1985) [inline app/uk/gov/hmrc/helptosave/connectors/DESConnector.scala:34]: does this not need to stay the same?

File: app/uk/gov/hmrc/helptosave/connectors/IFConnector.scala [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+import uk.gov.hmrc.helptosave.http.HttpClient.HttpClientOps
+import uk.gov.hmrc.helptosave.util.{Logging, NINO}
+import uk.gov.hmrc.http.{HeaderCarrier, HttpClient, HttpResponse}
+import uk.gov.hmrc.play.bootstrap.config.ServicesConfig
+
+import scala.concurrent.{ExecutionContext, Future}
+
+@ImplementedBy(classOf[IFConnectorImpl])
+trait IFConnector {
+  def ifCorrelationId(response: HttpResponse): String = response.header("IfCorrelationId").getOrElse("-")
```
Comment (johno1985) [inline app/uk/gov/hmrc/helptosave/connectors/IFConnector.scala:31]: this needs to be CorrelationId, as per the IF api spec

File: app/uk/gov/hmrc/helptosave/connectors/IFConnector.scala [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+  implicit val booleanShow: Show[Boolean] = Show.show(if (_) "Y" else "N")
+
+  val originatorIdHeader: (String, String) = "Originator-Id" -> servicesConfig.getString(
+    "microservice.services.paye-personal-details.originatorId")
+
+  def payePersonalDetailsUrl(nino: String): String = s"$payeURL$root/pay-as-you-earn/02.00.00/individuals/$nino"
+
+  override def getPersonalDetails(nino: NINO)(implicit hc: HeaderCarrier): Future[HttpResponse] =
+    http.get(payePersonalDetailsUrl(nino), headers = appConfig.ifHeaders + originatorIdHeader)(
+      hc.copy(authorization = None),
```
Comment (johno1985) [inline app/uk/gov/hmrc/helptosave/connectors/IFConnector.scala:52]: do we not need a token here?

File: app/uk/gov/hmrc/helptosave/util/HttpResponseOps.scala [hunk 10]
```diff
@@ -45,7 +45,9 @@ class HttpResponseOps(val response: HttpResponse) extends AnyVal {
       couldntParseJson = (_, e: JsError) => s"Could not parse http response JSON: ${e.prettyPrint()}"
     )
 
-  def desCorrelationId: String = response.header("CorrelationId").getOrElse("-")
+  def desCorrelationId: String = response.header("DesCorrelationId").getOrElse("-")
```
Comment (johno1985) [inline app/uk/gov/hmrc/helptosave/util/HttpResponseOps.scala:48]: I believe the correlation ids should both be "CorrelationId"

File: test/uk/gov/hmrc/helptosave/utils/TestSupport.scala [hunk 11]
```diff
@@ -75,17 +75,7 @@ trait TestSupport extends UnitSpec with IdiomaticMockito with BeforeAndAfterAll
 
   val servicesConfig: ServicesConfig = fakeApplication.injector.instanceOf[ServicesConfig]
 
-//  val mockMetrics: Metrics = new Metrics(stub[MetricRegistry]) {
```
Comment (johno1985) [inline test/uk/gov/hmrc/helptosave/utils/TestSupport.scala:78]: please delete commented code

**Review by johno1985 — CHANGES_REQUESTED** [review]

File: app/uk/gov/hmrc/helptosave/config/AppConfig.scala (hunk 6 above)
Comment (ruthdas03) [reply app/uk/gov/hmrc/helptosave/config/AppConfig.scala]: Removed it

File: app/uk/gov/hmrc/helptosave/connectors/DESConnector.scala (hunk 7 above)
Comment (ruthdas03) [reply app/uk/gov/hmrc/helptosave/connectors/DESConnector.scala]: Yes, made them same ie "CorrelationId" for DES and IF

File: app/uk/gov/hmrc/helptosave/connectors/IFConnector.scala (hunk 8 above)
Comment (ruthdas03) [reply app/uk/gov/hmrc/helptosave/connectors/IFConnector.scala]: Yes done

File: app/uk/gov/hmrc/helptosave/connectors/IFConnector.scala (hunk 9 above)
Comment (ruthdas03) [reply app/uk/gov/hmrc/helptosave/connectors/IFConnector.scala:50]: the token is set in ifHeaders in https://github.com/hmrc/help-to-save/blob/e39114ed40cb4a64ab31d1e18dc193d3d32dfdf2/app/uk/gov/hmrc/helptosave/config/AppConfig.scala#L44-L46

File: app/uk/gov/hmrc/helptosave/util/HttpResponseOps.scala (hunk 10 above)
Comment (ruthdas03) [reply app/uk/gov/hmrc/helptosave/util/HttpResponseOps.scala]: Yes done

File: app/uk/gov/hmrc/helptosave/connectors/IFConnector.scala [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+}
+
+@Singleton
+class IFConnectorImpl @Inject()(http: HttpClient, servicesConfig: ServicesConfig)(implicit appConfig: AppConfig, ec: ExecutionContext)
+    extends IFConnector with Logging {
+
+  val payeURL: String = servicesConfig.baseUrl("paye-personal-details-if")
+  val root: String = servicesConfig.getString("microservice.services.paye-personal-details-if.root")
+
+  implicit val booleanShow: Show[Boolean] = Show.show(if (_) "Y" else "N")
```
Comment (johno1985) [inline app/uk/gov/hmrc/helptosave/connectors/IFConnector.scala:41]: is this needed?

File: app/uk/gov/hmrc/helptosave/services/HelpToSaveService.scala [hunk 13]
```diff
@@ -133,58 +137,49 @@ class HelpToSaveServiceImpl @Inject()(
...
+      response.status match {
+          case Status.OK =>
+            response.parseJsonWithoutLoggingBody[PayePersonalDetails] tap {
+              case Left(e) =>
                 metrics.payePersonalDetailsErrorCounter.inc()
-                pagerDutyAlerting.alert("Received unexpected http status in response to paye-personal-details")
-                Left(s"Received unexpected status $other")
-
+                logger.warn(
+                  s"$tag Could not parse JSON response from paye-personal-details, received 200 (OK): $e ${timeString(time)}",
```
Comment (johno1985) [inline app/uk/gov/hmrc/helptosave/services/HelpToSaveService.scala:158]: what was the verdict on the tag? is it still needed?

File: conf/application.conf [hunk 14]
```diff
@@ -201,3 +214,5 @@ enrolment {
   delete-ninos = []
   undo-delete-ninos = []
 }
+
+feature.if.enabled = false
```
Comment (johno1985) [inline conf/application.conf:218]: should default to true

File: app/uk/gov/hmrc/helptosave/services/HelpToSaveService.scala (hunk 13 above)
Comment (ruthdas03) [reply app/uk/gov/hmrc/helptosave/services/HelpToSaveService.scala:158]: It's not needed in alert, I hope we can keep them in warn and debug levels

File: app/uk/gov/hmrc/helptosave/connectors/IFConnector.scala (hunk 12 above)
Comment (ruthdas03) [reply app/uk/gov/hmrc/helptosave/connectors/IFConnector.scala:41]: No, removed it

File: conf/application.conf (hunk 14 above)
Comment (ruthdas03) [reply conf/application.conf:218]: Done

**Review by LAJW — APPROVED** [review]

**Pull request merged**